### DISCUSSION TITLE
Rosetta stone work meeting notes

### DISCUSSION THREAD

**Discussion by kriskowal** [ocapn/ocapn#86] [General]
This is a summary of an ad-hoc work meeting with myself @kriskowal, @gibson042, @cwebber, dthompson, @tsyesika, and @michaelfig.

I’ve been luring various folks (@gibson042, @erights, @cwebber) to help fill out a [Rosetta stone for OCapN forms](https://docs.google.com/spreadsheets/d/11LZiqgI9syIHPhbDQE1fxMxGQSXHJGchT8JHbnf2wYo/edit#gid=0) between supported languages and concrete forms.

In this meeting, I received:

- Guile’s `*unspecified*`, JavaScript’s `undefined`, and Python’s `None` correspond generally,
  - …though there are cases where JavaScript’s `null` is a better fit for Python’s `None` (@cwebber please expand on that. I think even in kwarg defaults, `undefined` is still closer to `None` than `null`.)
- The current consensus is to keep a second bottom value, JavaScript’s `null`, which would be represented by an imported, singleton, unique identity, place-holder in Guile (tentatively named `zilch` to avoid confusion with elisp’s `null`) and some value in Python (I’m going to tentatively propose `from ocapn import Null` here.)
  - The assembled participants are willing to entertain avoiding a second bottom value, as this is consistent with making OCapN more language-neutral.
- The Syrup shorthand in the OCapN docs closely resembles Preserves.
  - In that shorthand, any nominal term is to be interpreted as a symbol (as in `<label>`) unless otherwise marked (as in `<"label>`, where the 0th field is a string instead).
  - This notation would ideally be either better documented or adopted _en mass_ from Preserves textual representation.
- OCapN as-implemented supports a `<user-record tag [payload]>` that loosely corresponds to Agoric’s user the tagged data underlying patterns and copy data like map, sets, and bags. Agoric’s form consists of a `tagName` that must be a `string` and arbitrary singular payload value.
- OCapN has a mechanism for communicating multi-return and we need to talk more about how that will or won’t manifest in JavaScript. The OCapN `<op:pick>` exists for this case and OCapN currently lacks an analog for Agoric’s `E.get(object).field` or `E.get(array)[index]`. (Aside: does Agoric’s notation oblige OCapN to allow `E.get(array).length` for lists?

In a future high-power work meeting:

- I would like to argue in favor of normalizing method invocation and
- eliminating symbols, in pursuit of making OCapN more language-neutral.
- Riff on muti-return envelope and what that looks like for `Far` objects in JavaScript.
- Riff on JavaScript `Far` object/actors that support both invocation and call.
- Propose a `<op:get receiver field-name>` for JavaScript’s `E.get(receiver).field`, which extracts JavaScript property values and discuss whether to also support indexes of lists or have a separate operation for that.
- Discuss the conflationg of nonce locator and bootstrap objects (which might be fine)

In a future low-power work meeting:

- I would like to learn more about Guile values and types so as to fill in the Guile columns more completely.
- Learn more about the final argument of `<op:deliver>`, which appears to oblige the caller to decide whether it’s asking for a promise or presence, and compare to the equivalent Agoric CapTP.
- What’s the method name again in deliver-op for applying a function instead of invoking a method? That’s `#f`, right?

**Comment by kriskowal** [comment]
@erights It occurs to me that `Actor` might be a clearer name than `Far` in the OCapN ecosystem.

**Reply by erights** [reply]
> @erights It occurs to me that `Actor` might be a clearer name than `Far` in the OCapN ecosystem.

The problem is that `Actor` as in the classic Hewitt / Baker / Clinger actor semantics corresponds best to a Vat as a whole. All correspondences are approx and could be shifted around, but I want to maintain the clarity of being able to explain our model's relationship to Actors as "Vats are multi-ported Actors". For example, a local synchronous call to a Far object can bring about other synchronous state change within the Vat. With "Vats are like Actors", we get to explain "Turns are like Events". IOW, Vats, like Actors, are only asynchronously coupled to each other. Far objects within a Vat can be as synchronously coupled to each other as any other two objects within the same Vat.

See chapter "From Objects to Actors and Back Again" in my thesis.

**Reply by erights** [reply]
But I agree that "Far object" is a terrible term, which is why we've been moving away from it.

**Comment by erights** [comment]
>   * The assembled participants are willing to entertain avoiding a second bottom value, as this is consistent with making OCapN more language-neutral.

I'll just note (as was already clear from @kriskowal 's note) that I was not among the assembled participants for this discussion today. FWIW, I hate having two bottom values. But as far as I'm concerned, in previous public threads on this repo, we already have entertained this and settled on two bottom values, for ugly but pragmatic reasons I've already explained. Is there any new info since those threads that would invalidate or weaken that previous reasoning?

The OCapN taxonomy of passable values, that we arrived at agreement on after much discussion (especially with @zenhack !) is best represented by https://github.com/ocapn/ocapn/issues/5#issuecomment-1549012122 and the thread descended from that message. Unless stated otherwise, I will continue to use the names suggested by Ian in that message. I am not attached to those names, but they represent a stable point from the agreements of our previous discussions, so let's stick to them until we agree to change them.

Three such changes I'd like us to consider (but will not switch to until we have agreement): 
- [ ] Rename `Sequence` to `List`?
- [ ] Rename `SignedInteger` to something as clear but shorter, like `Integer` (my favorite) or `BigInteger`?

----

My third suggestion is a deep semantic change, not just a change of terminology. As I mention in that thread, at Agoric, our `@endo/pass-style` level of abstraction is intended to (eventually) correspond exactly to the OCapN core data types taxonomy and semantics. There are several discrepancies we know about that we will fix on the Agoric side: lack of -0.0, lack of ByteString. But the open question for us has to do with the OCapN "Capability" category. For us, for what I think are compelling reasons, we distinguish between Promises and Remotables. From talking to @cwebber yesterday, before the meeting I missed today, it seems Spritely is aligned enough with the same splitting of the Capability category. (In Ian's absence, I don't know what to say about concerns from cap'n proto though.)

From @kriskowal 's notes about today's meeting, it seems like we're all still happy about that split, but that what we mean by Promise will need a lot more work to get to agreement.

Nevertheless, can we agree to 
- [ ] Split the current "Capability" category into "Remotable" and "Promise"?

I'll state the main difference between these two concepts as
* Indeed, they are both Capabilities, that carry and deliver messages to possibly remote objects
* Unlike Promises, Remotables are capabilities with an identity that is
   * stable
   * unforgeable
   * comparable
   * pass-invariant
   * (and hopefully adequate for grant matching)

In order to have the taxonomy itself remain pass-invariant, a Far object (or whatever we call it) is a remotable. Passing it results in a remote presence which is also a remotable. For the identity to be pass-invariant, passing the "same" remotable twice must arrive as the "same" remotable.

A passed promise must arrive as a promise and a passed remotable must arrive as a remotable with the same identity.

While postponing the remaining promise semantics to another argument, I'd like to see if we can get agreement on the above remotable semantics, and the pass-invariance of promise and remotable as distinct categories?

----

In case there are new reasons to drop one of the bottom types, or other reasons why we might arrive at joint agreement to do so, I'll go ahead and add that proposal to this message:
- [ ] Drop one bottom value (say "Null"), leaving only one remaining (say "Undefined").
(My current answer to this is no, for reasons previously stated.)

**Reply by kriskowal** [reply]
We were trying to arrive at a pair of bottom values for each language so that one was the favored version (like `undefined`, `*unspecified*`, and `None`) and the other was less favored (like `null`, `zilch`, and `Null`) so that we could herd conventions toward the more favored form. @cwebber wasn’t sure which way `None` leaned in Python. My feeling is that `None` more closely follows `undefined`. I suspect we have no new information. However, it would be fair to say that having a second bottom value to accommodate `null` is not consistent with the shared goal of language neutrality, on which I hope we base many design decisions, including single-return, normalization of method invocation, the obviation of passable symbols.

**Reply by erights** [reply]
Great! I agree on the general goal, and I am embarrassed that pragmatic considerations lead me to needing two bottoms. I acknowledge it is an ugly compromise, and I agree that we should minimize these.

Also, 100% agreement that it is worth distinguishing favored bottom from the other one. Also agreed that we need to favor the one that maps to the JS `undefined`, so I'm glad to stick with Ian's names for these two bottom values.

**Comment by erights** [comment]
> does Agoric’s notation oblige OCapN to allow E.get(array).length for lists?

s/lists/Sequences (or hopefully Lists if we agree on that renaming).

I don't understand the question. How could OCapN disallow endpoint logic from locally supporting `E.get(sequence).length` ?
Perhaps you're asking whether OCapN should support the pipelining of such queries about remote data? That is interesting. AFAIK, Agoric does not yet. But it would be nice to be able to.

**Reply by erights** [reply]
> Propose a <op:get receiver field-name> for JavaScript’s E.get(receiver).field, which extracts JavaScript property values and discuss whether to also support indexes of lists or have a separate operation for that.

Gotcha. So the question is whether OCapN should support an `op:get` that includes asking a Sequence for its length? I agree that would be attractive. I'd also like to ask a Struct for its property names.

**Comment by kriskowal** [comment]
@cwebber So, we’ve learned in this thread that the purpose for “other bottom” is to round-trip JSON on top of OCapN. Being able to round-trip arbitrary JSON would probably be useful to Guile too. So, two observations:

1. `<'jsonnull>` is a reasonable representation for the other bottom in the Syrup notational variant, and `jsonnull` is a reasonable name for the corresponding value in Guile. But,
2. I see that other Guile JSON libraries use the symbol `'null`. If we went with that, it would be possible to parse some JSON and send it over OCapN from Guile, receive it and write JSON in JavaScript, without loss. For this to be possible, we would have to forgo representing symbols as values on the wire, which I hope to convince you would be valuable anyway, just for advancing language neutrality.
